src/qt/window.cpp

```diff
@@ -353,6 +361,16 @@ void wxWindowQt::PostCreation(bool generic)
     HandleWindowEvent(event);
 }
 
+bool wxWindowQt::Destroy() // do not execute wxWindowBase version which would immediately delete this
```

Comment:
But how is the window actually going to be destroyed if we don't do this? Sorry if I'm missing something here, but it looks like with this change calling `Destroy()` is actually never going to destroy the window at all (well, at least until its parent is destroyed normally and deletes it too).

Reply:
Interesting question. I always thought that `Destroy()` would postpone deletion until idle, when some magic will happen and pending-deletion windows will actually be deleted (a loop based on `IsBeingDestroyed()` or something like that).
So you mean there is no such thing? And normally when `Destroy()` is called all windows are actually deleted unless there is an override? Because if that is true, then you are right, child windows would not get deleted (before their parent), and probably a `DeletePendingDestructionChildWindows()` should be added to be run at idle?

Reply:
This is true for TLWs, see `wxTopLevelWindowBase::Destroy()`, but child windows are destroyed immediately. In principle, as we were always careful to state that this is not guaranteed, wxQt could delay deleting them too but, first, this would actually need to be done, unlike now, and, second, I'm afraid this will still result in problems due to assumptions in the existing code, so it'd be better to avoid this.

For now I'm closing this one as it can't be applied if I didn't miss anything and if it really doesn't delete the windows at all.

Reply:
@vadz This could be split into 2 parts:
- The changes to `~wxWindowQt()`, which should still be applied even without the `Destroy()` override (well, in this case without the check for `IsBeingDeleted()`);
- A way for having delayed destruction.

For the second part: `Qt` does have `deleteLater()` functionality, which will keep a targeted `QObject` until idle and only then delete it. But this does not look like a good approach, because calling it for a `QObject` owned by a `wxWindow` will mean (even if mostly theoretically) that each one will need to live for a while without relying on the other to still be alive.
Do you think any of the following possible solutions could work?

1. in `wxWindowQt::Destroy()`, add `this` to a list of windows that are pending deletion;
   at idle, the app instance could iterate through its top level windows and call a new `DeletePendingWindows()` and so on.
2. in `wxWindowQt::Destroy()` add a call that will execute at idle and which will only `delete this`, i.e. `CallAfter(&wxWindowQt::DeleteIfPending)`.

Reply:
Do we actually need to have deleted destruction for non-TLWs in wxQt? I.e. can't we just flush all the events for them and destroy them immediately?

If yes, then I think both of the proposed solutions would work, but, as I wrote, this risks breaking user code which relies (even though it shouldn't...) on the window being destroyed immediately (and so its dtor being executed, for example).

Reply:
> Do we actually need to have deleted destruction for non-TLWs in wxQt?

_delayed_ destruction?
Yes. For example without it I can see crashes related to Qt focus events.

> I.e. can't we just flush all the events for them and destroy them immediately?

I guess that would work too, i.e. calling `QCoreApplication::removePostedEvents` _and_ `blockSignals(true)` in `Destroy` as well, but IMHO that looks rather ugly...

I've just tested something along the lines of `2.` above (in the current implementation, added `CallAfter` with a new function that only deletes `this`), made sure each window gets deleted and it doesn't seem to create any problem.

> this risks breaking user code which relies (even though it shouldn't...) on the window being destroyed immediately (and so its dtor being executed, for example).

I've just re-read the docs of `wxWindow::Destroy()`. They explicitly say that owned windows have delayed destruction and imply (?) that the others don't. The `Window Deletion` overview does clearly state it, but even that is one of the last phrases of a long text that always talks about "windows"... Personally in 10+ years of wxW I never managed to remember this difference between managed windows and the others - I tried hard to use `Destroy()` exactly because what I thought it did.
Anyway, if you think that immediate deletion should be enforced (which is quite sad IMHO) then `Destroy()` must contain those calls. But it definitely cannot stay like this as it will eventually lead to crashes.

Reply:
@vadz Any opinion on the last comment? (No actual hurry but I don't know if you still receive notifications about comments in closed PRs.)

Reply:
Sorry, I did see it, but I'm not sure about this. My first instinct would remain to try to delete the window immediately as even though I don't have any concrete examples (but then how could I, as so far we didn't have any port delaying child windows destruction), I'm 99% sure it will create problems. E.g. consider destroying a child window, creating another one and relaying out the containing sizer: if you don't really destroy it, the layout will be broken. Of course, you could work around this particular problem by hiding the window immediately and hope it wasn't using `wxRESERVE_SPACE_EVEN_IF_HIDDEN` style, but it's just a single example, there are certainly others.

So I'd still go with `removePostedEvents()` approach, even though it's not as theoretically pure as the other one. But IMO it will create fewer problems in practice.

Another possibility would be to detach the `QWindow` and destroy it later, even though `wxWindow` would be deleted immediately. Would this result in any problems?

Reply:
@vadz 
> Another possibility would be to detach the `QWindow` and destroy it later, even though `wxWindow` would be deleted immediately. Would this result in any problems?

I suspect it would. Current code does not always rely on a single pointer to a qt object in a wx wrapper (i.e. `QWidget *m_qtWindow` in `wxWindowQt` and `QRadioButton *m_qtRadioButton` in `wxRadioButton` point to the same thing). Then there is all the code that does not check the qt member for NULL, and the code in the qt-derived object that does not check the wx counterpart (the "handle**r**") for NULL. I cannot promise I'll be able to chase all those cases and make sure they are safe. There is also the problem of events and signals that are emitted by the qt object while it is pending deletion but after it was detached.
Detaching the 2 entities does not sound very appealing unfortunately.

I'll make the changes for immediate destruction in `Destroy()`. Should I update (and reopen? is that possible?) this PR or create a new one?

Reply:
> @vadz
> 
> > Another possibility would be to detach the `QWindow` and destroy it later, even though `wxWindow` would be deleted immediately. Would this result in any problems?
> 
> I suspect it would. Current code does not always rely on a single pointer to a qt object in a wx wrapper (i.e. `QWidget *m_qtWindow` in `wxWindowQt` and `QRadioButton *m_qtRadioButton` in `wxRadioButton` point to the same thing). 

But if the wx object is destroyed anyhow, why would any pointers stored in it matter?

>Then there is all the code that does not check the qt member for NULL, and the code in the qt-derived object that does not check the wx counterpart (the "handle**r**") for NULL. 

The latter one is indeed bad. We would have to use `wxWeakRef<>`s everywhere to fix it safely. Probably not worth it...

> I'll make the changes for immediate destruction in `Destroy()`. Should I update (and reopen? is that possible?) this PR or create a new one?

I think it would be simpler to just create a new one but I can also reopen this one if you prefer (I don't know if you can do it or not to be honest). TIA!